src/charm.py

```diff
@@ -302,20 +369,212 @@ def _on_admin_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
+        )
+
+    def _on_client_changed(self, event: ClientChangedEvent) -> None:
+        if not self._container.can_connect():
+            event.defer()
+            return
+
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            event.defer()
+            return
+
+        client_config = event.to_client_config()
+        try:
+            self._hydra_cli.update_client(
+                client_config, metadata=f'{{"relation_id": {event.relation_id}}}'
+            )
+        except ExecError as err:
+            logger.error(f"Exited with code: {err.exit_code}. Stderr: {err.stderr}")
+            return
+        except Error as e:
+            logger.error(f"Something went wrong when trying to run the command: {e}")
+            logger.info("Deferring the event")
+            event.defer()
+            return
+
+    def _on_client_deleted(self, event: ClientDeletedEvent) -> None:
+        if not self._container.can_connect():
+            event.defer()
+            return
+
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            event.defer()
+            return
+
+        client = self._get_oauth_relation_peer_data(event.relation_id)
+        if not client:
+            logger.error("No client found in peer data")
+
+        try:
+            self._hydra_cli.delete_client(client["client_id"])
+        except ExecError as err:
+            logger.error(f"Exited with code: {err.exit_code}. Stderr: {err.stderr}")
+            return
+        except Error as e:
+            logger.error(f"Something went wrong when trying to run the command: {e}")
+            logger.info("Deferring the event")
+            event.defer()
+            return
+
+        self._pop_oauth_relation_peer_data(event.relation_id)
+
+    def _update_endpoint_info(self) -> None:
+        if not self.admin_ingress.url or not self.public_ingress.url:
+            return
+
+        self.oauth.set_provider_info_in_relation_data(
+            issuer_url=self.public_ingress.url,
+            authorization_endpoint=join(self.public_ingress.url, "oauth2/auth"),
+            token_endpoint=join(self.public_ingress.url, "oauth2/token"),
+            introspection_endpoint=join(self.admin_ingress.url, "admin/oauth2/introspect"),
+            userinfo_endpoint=join(self.public_ingress.url, "userinfo"),
+            jwks_endpoint=join(self.public_ingress.url, ".well-known/jwks.json"),
+            scope=SUPPORTED_SCOPES,
+        )
+
+
+class HydraCLI:
+    """Helper object for running hydra CLI commands."""
+
+    def __init__(self, hydra_admin_url: str, container: Container):
+        self.hydra_admin_url = hydra_admin_url
+        self.container = container
+
+    def _client_config_to_cmd(
+        self, client_config: ClientConfig, metadata: Optional[Union[Dict, str]] = None
```

Comment:
Is there a reason why we want to accept both string and dit for the metadata? Seems simpler to settle on one